WebGPU: Compute passes

mwyrzykowski opened the pull request:
#### af82bfdea3946b9862c5291ac5ef1c301c31e5e4
<pre>
WebGPU: Compute passes
<a href="https://bugs.webkit.org/show_bug.cgi?id=250863">https://bugs.webkit.org/show_bug.cgi?id=250863</a>
&lt;radar://83213548&gt;

Reviewed by Myles C. Maxfield.

Implement compute passes which are needed to start
passing a lot of the CTS tests which create a simple
compute shader and read back the result, to ensure
the adapter was created.

<a href="https://bugs.webkit.org/show_bug.cgi?id=251171">https://bugs.webkit.org/show_bug.cgi?id=251171</a> has
been filed for tracking getting the threads per thread
group size from the shader compiler.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::beginRenderPass):

* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::dispatch):
(WebGPU::ComputePassEncoder::dispatchIndirect):
(WebGPU::ComputePassEncoder::setBindGroup):
(WebGPU::ComputePassEncoder::setPipeline):
(WebGPU::ComputePassEncoder::endPass):

* Source/WebGPU/WebGPU/BindGroupLayout.h:
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::BindGroupLayout::createEntryFromStructMember):
Moved so both compute and render pipeline can access.

* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
(WebGPU::ComputePipeline::threadsPerThreadgroup const):

* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::buildKeyValueReplacements const):
Moved to common path.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
(WebGPU::buildKeyValueReplacements): Deleted.
(WebGPU::createEntryFromStructMember): Deleted.
Moved to common path.

Canonical link: <a href="https://commits.webkit.org/259634@main">https://commits.webkit.org/259634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/838a0cc0dd4d61ad158eeacb194ecc706ce1e93a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105477 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114735 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15871 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/5797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97781 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12161 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39653 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94026 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7864 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/8217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/5797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13991 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/47684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6648 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/9889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -116,8 +116,16 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
+        return ComputePassEncoder::createInvalid(m_device);
+
+    MTLComputePassDescriptor* computePassDescriptor = [MTLComputePassDescriptor computePassDescriptor];
+    computePassDescriptor.dispatchType = MTLDispatchTypeSerial;
```
Cool, I think this is right

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -116,8 +116,16 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
+    MTLComputePassDescriptor* computePassDescriptor = [MTLComputePassDescriptor computePassDescriptor];
+    computePassDescriptor.dispatchType = MTLDispatchTypeSerial;
+
+    id<MTLComputeCommandEncoder> computeCommandEncoder = [m_commandBuffer computeCommandEncoderWithDescriptor:computePassDescriptor];
```
I'm glad to see us using descriptors here, I think that's the right direction

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -131,7 +139,6 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
 
 Ref<RenderPassEncoder> CommandEncoder::beginRenderPass(const WGPURenderPassDescriptor& descriptor)
 {
-    UNUSED_PARAM(descriptor);
```
👍

litherum commented on Source/WebGPU/WebGPU/ComputePassEncoder.h:
```diff
@@ -85,6 +85,8 @@ class ComputePassEncoder : public WGPUComputePassEncoderImpl, public RefCounted<
     uint64_t m_debugGroupStackSize { 0 };
 
     const Ref<Device> m_device;
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=251171 - this should come from the WGSL compiler
```
Ok

litherum commented on Source/WebGPU/WebGPU/ComputePassEncoder.mm:
```diff
@@ -55,15 +55,16 @@
+    if ((x % m_threadsPerThreadgroup.width) || (y % m_threadsPerThreadgroup.height) || (z % m_threadsPerThreadgroup.depth))
+        return;
+
+    [m_computeCommandEncoder dispatchThreads:MTLSizeMake(x / m_threadsPerThreadgroup.width, y / m_threadsPerThreadgroup.height, z / m_threadsPerThreadgroup.depth) threadsPerThreadgroup:m_threadsPerThreadgroup];
```
I think this is wrong for 2 reasons:

1. DispatchThreads isn't available on all devices IIRC
2. I believe the arguments indicate the number of workgroups to launch, not the number of threads to launch

mwyrzykowski replied:
Great catch! Totally meant to write `dispatchThreadgroups`

litherum replied:
Please fixup (or, ideally, remove) the math above this, to make it appropriate for dispatchThreadgroups

litherum commented on Source/WebGPU/WebGPU/ComputePassEncoder.mm:
```diff
@@ -55,15 +55,16 @@
 {
-    UNUSED_PARAM(indirectBuffer);
-    UNUSED_PARAM(indirectOffset);
+    [m_computeCommandEncoder dispatchThreadgroupsWithIndirectBuffer:indirectBuffer.buffer() indirectBufferOffset:indirectOffset threadsPerThreadgroup:m_threadsPerThreadgroup];
```
FIXME: Validate this

mwyrzykowski committed:
WebGPU: Compute passes
https://bugs.webkit.org/show_bug.cgi?id=250863
<radar://83213548>

Reviewed by Myles C. Maxfield.

Implement compute passes which are needed to start
passing a lot of the CTS tests which create a simple
compute shader and read back the result, to ensure
the adapter was created.

https://bugs.webkit.org/show_bug.cgi?id=251171 has
been filed for tracking getting the threads per thread
group size from the shader compiler.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::beginRenderPass):

* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::dispatch):
(WebGPU::ComputePassEncoder::dispatchIndirect):
(WebGPU::ComputePassEncoder::setBindGroup):
(WebGPU::ComputePassEncoder::setPipeline):
(WebGPU::ComputePassEncoder::endPass):

* Source/WebGPU/WebGPU/BindGroupLayout.h:
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::BindGroupLayout::createEntryFromStructMember):
Moved so both compute and render pipeline can access.

* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
(WebGPU::ComputePipeline::threadsPerThreadgroup const):

* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::buildKeyValueReplacements const):
Moved to common path.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
(WebGPU::buildKeyValueReplacements): Deleted.
(WebGPU::createEntryFromStructMember): Deleted.
Moved to common path.

Canonical link: https://commits.webkit.org/259634@main
